### PR TITLE
ledger snapshots

### PR DESCRIPTION
**Pull request by evanmcc**
miner support and tests for:

https://github.com/helium/blockchain-core/pull/403

### CODE REVIEW

File: src/miner.erl [hunk 1]
```diff
@@ -367,6 +369,31 @@ handle_call({create_block, Metadata, Txns, HBBFTRound}, _From, State) ->
...
                     Metadata),
+    %% TODO what do we do about snapshot disagreements?
+    %% IIRC we had agreed to omit
+    SnapshotHash=
+        case blockchain:config(?snapshot_interval, blockchain:ledger(Chain)) of
+            {ok, Interval} ->
+                case (NewHeight - 1) rem Interval == 0 of
+                    true ->
+                        lager:info("XXX trying to add hash ~p", [NewHeight]),
+                        [SH] = lists:foldl(
```
Comment (Vagabond) [inline src/miner.erl:380]: badmatching here would presumably halt block production?

Comment (evanmcc) [reply src/miner.erl:380]: yes, I need to change this to 2f+1 agreement, and omission if it does not meet that standard.

File: rebar.config [hunk 2]
```diff
@@ -11,7 +11,8 @@
 {deps, [
         {blockchain, {git, "https://github.com/helium/blockchain-core.git",
                       {branch, "master"}}},
-        {hbbft, {git, "https://github.com/helium/erlang-hbbft.git", {branch, "master"}}},
+        {hbbft, {git, "https://github.com/helium/erlang-hbbft.git",
```
Comment (Vagabond) [inline rebar.config:14]: This branch merged.

Comment (evanmcc) [reply rebar.config:14]: fixed

File: src/miner.erl [hunk 3]
```diff
@@ -369,6 +371,35 @@ handle_call({create_block, Metadata, Txns, HBBFTRound}, _From, State) ->
...
+                        SHCt =
+                            lists:foldl(
+                              fun({_Idx, #{snapshot_hash := SH}}, Acc) -> % new map vsn
+                                      maps:update_with(SH, fun(V) -> V + 1 end, 1, Acc);
+                                 (_, Acc) ->
+                                      Acc
+                              end,
+                              #{},
+                              Metadata),
+                        case lists:reverse(lists:keysort(2, maps:to_list(SHCt))) of
```
Comment (Vagabond) [inline src/miner.erl:394]: So we're sorting the most common ones first?

Comment (Vagabond) [reply src/miner.erl:394]: Couple comments here would be good, I think.

File: src/miner.erl [hunk 4]
```diff
@@ -369,6 +371,35 @@ handle_call({create_block, Metadata, Txns, HBBFTRound}, _From, State) ->
                     end,
                     [],
                     Metadata),
+    %% TODO what do we do about snapshot disagreements?
```
Comment (Vagabond) [inline src/miner.erl:374]: Update this now we have decided

File: src/miner.erl [hunk 5]
```diff
@@ -414,7 +451,8 @@ handle_call({create_block, Metadata, Txns, HBBFTRound}, _From, State) ->
                                election_epoch => ElectionEpoch,
                                epoch_start => EpochStart,
                                seen_votes => SeenVectors,
-                               bba_completion => BBA
+                               bba_completion => BBA,
+                               snapshot_hash => SnapshotHash
```
Comment (Vagabond) [inline src/miner.erl:455]: The empty binary is the PBC default value, right?

File: src/miner_hbbft_sidecar.erl [hunk 6]
```diff
@@ -74,9 +74,16 @@ init(State) ->
             erlang:send_after(500, self(), chain_check),
             {ok, State#state{}};
         Chain ->
-            Ledger = blockchain_ledger_v1:new_context(blockchain:ledger(Chain)),
-            Chain1 = blockchain:ledger(Ledger, Chain),
-            {ok, State#state{chain = Chain1}}
+            Ledger0 = blockchain:ledger(Chain),
+            case Ledger0 of
```
Comment (Vagabond) [inline src/miner_hbbft_sidecar.erl:78]: Do we still have the undefined ledger so we need this?

File: src/miner.erl (hunk 3 above)
Comment (evanmcc) [reply src/miner.erl:394]: added

File: src/miner.erl (hunk 5 above)
Comment (evanmcc) [reply src/miner.erl:455]: yes, this should cause nothing to be put into the proto for a bytes field, but we should make sure that a mixed cluster can make progress.

File: test/miner_blockchain_SUITE.erl [hunk 7]
```diff
@@ -26,7 +26,9 @@ all() -> [
           group_change_test,
           master_key_test,
           version_change_test,
-          election_v3_test
+          election_v3_test,
+          snapshot_test,
+          high_snapshot_test
```
Comment (Vagabond) [inline test/miner_blockchain_SUITE.erl:31]: Do we need a test where the snapshot is fixing a corrupted ledger?

File: src/miner.erl (hunk 5 above)
Comment (Vagabond) [reply src/miner.erl:463]: This has been confirmed

**Pull request merged**